{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6366e2de",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import numpy as np\n",
    "from pandas import DataFrame\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import pathlib\n",
    "import gzip\n",
    "import itertools\n",
    "import random\n",
    "from collections import namedtuple\n",
    "from collections import Counter\n",
    "from scipy import stats\n",
    "from scipy.stats import norm\n",
    "\n",
    "%matplotlib inline\n",
    "plt.rcParams['figure.figsize'] = [16.5, 5]\n",
    "plt.rcParams['font.size'] = 12\n",
    "plt.rcParams['pdf.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de3499c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da98828",
   "metadata": {},
   "source": [
    "### Define paths to input files and analysis parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1557669b",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = \"/pbi/flash/tmokveld/tmp/egor_ashg_analysis/update/merge.vcf.gz\" # A VCF generated by `trgt merge`\n",
    "min_allele_mult = 5  # Minimum allele multiplicity (remove rare alleles)\n",
    "min_alleles_per_tr = 500  # TRs with too few alleles will be removed\n",
    "min_cpgs_to_call_meth = 5 # Ignore allele methylation if it has fewer CpGs than this\n",
    "min_alleles_with_meth = 250 # Ignore TR methylation if methylation is called in fewer alleles than this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdc1038",
   "metadata": {},
   "source": [
    "Prepare output paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f387e602",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathlib.Path(\"output/\").mkdir(exist_ok=True)\n",
    "pathlib.Path(\"scratch/\").mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26435b86",
   "metadata": {},
   "source": [
    "### Summarize merged VCF into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7e3fd979",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_allele_recs(tr_rec):\n",
    "    all_alleles = []\n",
    "    for alleles in tr_rec.gts.values():\n",
    "        all_alleles.extend(alleles)\n",
    "    return all_alleles\n",
    "\n",
    "\n",
    "def num_samples_with_gt(tr_rec):\n",
    "    return len([als for als in rec.gts.values() if als])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36bd477c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_num_distinct_alleles(alleles):\n",
    "    score = len(set(a.seq for a in alleles))\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "411e1cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_frequent_alleles(alleles, min_allele_mult):\n",
    "    seqs = [a.seq for a in alleles]\n",
    "    freq_seqs = set(seq for seq, count in Counter(seqs).items() if count >= min_allele_mult)\n",
    "    return [a for a in alleles if a.seq in freq_seqs]\n",
    "\n",
    "\n",
    "def get_top_seq(alleles):\n",
    "    seqs = [a.seq for a in alleles]\n",
    "    counted_seqs = list(Counter(seqs).items())\n",
    "    counted_seqs.sort(key=lambda rec: rec[1], reverse=True)\n",
    "    return counted_seqs[0][0]\n",
    "\n",
    "\n",
    "def get_median_meth(alleles):\n",
    "    meths = [a.meth for a in alleles if a.meth is not None and a.seq.count(\"CG\") >= min_cpgs_to_call_meth]\n",
    "    if len(meths) >= min_alleles_with_meth:\n",
    "        return np.median(meths), np.std(meths)\n",
    "    else:\n",
    "        return None, None\n",
    "\n",
    "    \n",
    "def get_median_len(alleles):\n",
    "    lens = [len(a.seq) for a in alleles]\n",
    "    return np.median(lens), np.std(lens)\n",
    "\n",
    "\n",
    "def get_tr_type(rec):\n",
    "    if rec.trid.count(\",\") > 0:\n",
    "        return \"vc\"\n",
    "    \n",
    "    fields = rec.trid.split(\"-\")\n",
    "    if len(fields) != 4:\n",
    "        return None\n",
    "    chrom, start, end, motif = fields\n",
    "    start, end = int(start), int(end)\n",
    "\n",
    "    left_extension = abs(rec.locus.start - start)\n",
    "    right_extension = abs(rec.locus.end - end)\n",
    "    if left_extension > 5 or right_extension > 5:\n",
    "        return \"vc\"\n",
    "    \n",
    "    return \"solo\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e7448f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode(value):\n",
    "    if value is None:\n",
    "        return \"NA\"\n",
    "    if type(value) == int:\n",
    "        return str(value)\n",
    "    if type(value) in [float, np.float64]:\n",
    "        return f\"{value:.2f}\"\n",
    "    assert False, f\"Unexpected value '{value}' ({type(value)})\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c678f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping VWA1\n",
      "Skipping DAB1\n",
      "Skipping ABCD3\n",
      "Skipping NOTCH2NLA\n",
      "Skipping NOTCH2NLC\n",
      "Skipping NUTM2B-AS1\n",
      "Skipping C11ORF80\n",
      "Skipping CBL\n",
      "Skipping ATN1\n",
      "Skipping DIP2B\n",
      "Skipping ATXN2\n",
      "Skipping RILPL1\n",
      "Skipping ZIC2\n",
      "Skipping PABPN1\n"
     ]
    }
   ],
   "source": [
    "num_skipped = 0\n",
    "\n",
    "with open(\"output/summary_table.tsv\", \"w\") as outfile:\n",
    "    for index, rec in enumerate(vcf.get_rec(vcf_path)):\n",
    "        alleles = get_allele_recs(rec)\n",
    "        alleles = get_frequent_alleles(alleles, min_allele_mult)\n",
    "        if len(alleles) < min_alleles_per_tr:\n",
    "            num_skipped += 1\n",
    "            continue\n",
    "\n",
    "        tr_type = get_tr_type(rec)\n",
    "        if tr_type is None:\n",
    "            print(f\"Skipping {rec.trid}\")\n",
    "            continue\n",
    "\n",
    "        top_seq = get_top_seq(alleles)\n",
    "        len_median, len_std = get_median_len(alleles)\n",
    "        num_alleles = len(alleles)\n",
    "        num_distinct_alleles = get_num_distinct_alleles(alleles)\n",
    "        meth_median, meth_std = get_median_meth(alleles)\n",
    "    \n",
    "        fields = [rec.trid, tr_type, top_seq, str(num_alleles), str(num_distinct_alleles),\n",
    "                 encode(len_median), encode(len_std), encode(meth_median), encode(meth_std)]\n",
    "        print(\"\\t\".join(fields), file=outfile)\n",
    "\n",
    "\n",
    "\n",
    "print(f\"TRs processed: {index + 1}\")\n",
    "print(f\"Number of TRs skipped: {num_skipped} ({100 * num_skipped / (index + 1):.2f}%)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87dba080",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e348f55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3754abee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e64f32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
